hpke/shortkem.go

```diff
@@ -49,122 +68,84 @@ func (s shortKEM) DeriveKeyPair(seed []byte) (kem.PublicKey, kem.PrivateKey) {
 	}
 
 	bitmask := byte(0xFF)
-	if s.Params().BitSize == 521 {
+	if s.Curve == ecdh.P521() {
 		bitmask = 0x01
 	}
 
 	dkpPrk := s.labeledExtract([]byte(""), []byte("dkp_prk"), seed)
-	var bytes []byte
-	ctr := 0
-	for skBig := new(big.Int); skBig.Sign() == 0 || skBig.Cmp(s.Params().N) >= 0; ctr++ {
-		if ctr > 255 {
-			panic("derive key error")
-		}
-		bytes = s.labeledExpand(
+	for ctr := 0; ctr <= 255; ctr++ {
+		bytes := s.labeledExpand(
 			dkpPrk,
 			[]byte("candidate"),
 			[]byte{byte(ctr)},
 			uint16(s.byteSize()),
 		)
 		bytes[0] &= bitmask
-		skBig.SetBytes(bytes)
+		sk, err := s.UnmarshalBinaryPrivateKey(bytes)
```

Comment:
I confirmed that [Curve.NewPrivateKey](https://pkg.go.dev/crypto/ecdh#Curve) (called by UnmarshalBinaryPrivateKey) checks that the value is lower than the order of the curve, so is equivalent to the existing check.